feat(table-viz): rewrite with react-table and add query mode switch mode

ktmud opened the pull request:
### SUMMARY

This PR combines two major updates for the table chart:

1. Replace its rendering from `jquery.DataTables` to `react-table`: apache-superset/superset-ui#623
2. Rearrange the control panel, replace GROUP BY and NOT GROUP BY with a "Query Mode" switch: apache-superset/superset-ui#609 . It was not super clear what do GROUP BY and NOT GROUP BY mean and that they are mutually exclusive. Adding a tab switch should make things a little bit easier to understand.

In addition to these, there are also some smaller refactoring and UX improvements:

1. Clean up the legacy backend API for TableViz, add support for a new `query_mode` field.
2. Refactor control panel config management
3. Change default customize options for table chart
   - Time format now defaults to smartDate, which formats the time column based on time grain
   - Page length now defaults to NULL, which automatically add pagination of 200 rows per page when number of cells > 10,000

Must wait for release of new versions of `@superset-ui/chart-controls` and `@superset-ui/plugin-chart-table`.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

#### Before

<img src="https://user-images.githubusercontent.com/335541/85184188-a7716180-b243-11ea-83c9-0c449427b065.png" width="300">

#### After 
<img src="https://user-images.githubusercontent.com/335541/84823553-6410bc00-afd3-11ea-8148-3816702cd844.png" width="300">


### TEST PLAN

Manual testing

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [x] Has associated issue:
- [x] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

ktmud committed:
feat: upgrade react-table and table viz

Also optimize hot loader for ChartRenderer

ktmud committed:
feat(viz): add query mode switch to table chart

ktmud commented on superset/viz.py:
```diff
@@ -563,6 +567,46 @@ class TableViz(BaseViz):
+                - Otherwise defaults to AGG mode
+           2. Determine output columns based on query mode.
+        """
+        super().process_metrics()
```
Refactor how data fields are processed for `TableViz`. In general, for all visualizations, we should process all inputs as soon as possible, store the parsed params somewhere, and refer the single source of truth in all future operations.

kristw reviewed (approved):
approve the js part

ktmud commented on superset-frontend/src/explore/components/ControlPanelsContainer.jsx:
```diff
@@ -47,42 +48,11 @@ class ControlPanelsContainer extends React.Component {
-      };
-    }
-    return control;
-  }
```
All these logics are now handled by `expandControlConfig` from `superset-ui/chart-controls`.

ktmud commented on tests/viz_tests.py:
```diff
@@ -344,33 +344,28 @@ def test_adhoc_filters_overwrite_legacy_filters(self):
-        super_query_obj.return_value = {}
-        test_viz = viz.TableViz(datasource, form_data)
         with self.assertRaises(Exception):
+            test_viz = viz.TableViz(datasource, form_data)
```
The exception now raises earlier.

ktmud commented on tests/viz_tests.py:
```diff
@@ -390,21 +385,23 @@ def test_query_obj_merges_all_columns(self, super_query_obj):
-        self.assertEqual(["colA", "colB", "__time__"], query_obj["metrics"])
-        self.assertEqual([("__time__", True)], query_obj["orderby"])
+        self.assertEqual(["colA", "colB", "sort_column"], query_obj["metrics"])
+        self.assertEqual([("sort_column", True)], query_obj["orderby"])
```
Rename so not to be confused with the time column (`__timestamp`).

ktmud reviewed (commented):
@john-bodley @michellethomas would you mind having a look on the python part?

ktmud commented on superset-frontend/spec/javascripts/explore/exploreActions_spec.js:
```diff
@@ -22,13 +22,13 @@ import exploreReducer from 'src/explore/reducers/exploreReducer';
-    expect(newState.controls.x_axis_label.value).toBe('x');
-    expect(newState.form_data.x_axis_label).toBe('x');
+    expect(newState.controls.NEW_FIELD.value).toBe('x');
+    expect(newState.form_data.NEW_FIELD).toBe('x');
```
`x_axis_label` is not an active control at initialization or a shared control in global state. Use an ugly name to avoid confusion.

villebro commented on superset/viz.py:
```diff
@@ -553,6 +554,10 @@ def json_data(self) -> str:
         return json.dumps(self.data)
 
 
+QUERY_MODE_RAW = "raw"
+QUERY_MODE_AGGREGATE = "aggregate"
```
These should probably be placed in `utils/core.py` as
```python
class QueryMode(str, Enum):
    RAW = "raw"
   AGGREGATE = "aggregate"
```
I would also suggest adding a deprecated type for detecting query mode, which would look at the presence of `all_columns`, and flag as `raw` is present, and `aggregate` if not.

ktmud replied:
Yeah, I was unsure about how to do string enum in Python, so went with the easiest route.

Should we have a `core/consts.py` file?

ktmud replied:
> I would also suggest adding a deprecated type for detecting query mode (`AUTO`?), which would look at the presence of `all_columns`, and flag as `raw` is present, and `aggregate` if not.

The logic is there, it's just not its own query mode. Kind of feel it'd be more useful to add a `QueryMode.from_str(str)` method than introducing an `auto` mode.

villebro replied:
> Yeah, I was unsure about how to do string enum in Python, so went with the easiest route.
> 
> Should we have a `core/consts.py` file?

Oh yes, it's been put off for too long. I think it's more pythonic to call it `constants.py`. @john-bodley any strong feelings of how we could consolidate constants, enums and perhaps type aliases in one easy to find place? Having constants under `utils/core.py` is slightly confusing.

Anyway, I suggest not doing that refactor in this PR, let's do that separately. So I propose just throwing this under `utils/core.py` for now.

ktmud replied:
There is a `superset/contants.py` file, but nothing much seems going on there.

Just added a `QueryMode` in `utils/core.py` as well as a `Enum.get` method to safeguard invalid enum values.

villebro reviewed (commented):
Quick first pass. We probably should add `query_mode` into `QueryObject` both here and on `@superset-ui/query`, and use that as the primary means of either grouping or not grouping a query. The plan in SIP-38 is to deprecate the `groupby` property and only use `columns`, and then perform grouping only if grouping is requested (we also hope to be introducing non-aggregating metrics soon)

ktmud commented:
> We probably should add `query_mode` into `QueryObject` both here and on `@superset-ui/query`

Do you mind if I do this in a separate PR?

villebro commented:
> > We probably should add `query_mode` into `QueryObject` both here and on `@superset-ui/query`
> 
> Do you mind if I do this in a separate PR?

@ktmud sure thing, just wanted to put the comments here to make sure I don't forget mentioning it. I can also do it, I have an idea of how we can do it with minimal disruption.

ktmud commented on superset-frontend/src/explore/components/Control.jsx:
```diff
@@ -27,7 +27,10 @@ const controlTypes = Object.keys(controlMap);
   actions: PropTypes.object.isRequired,
   name: PropTypes.string.isRequired,
-  type: PropTypes.oneOf(controlTypes).isRequired,
+  type: PropTypes.oneOfType([
+    PropTypes.oneOf(controlTypes).isRequired,
+    PropTypes.func.isRequired,
+  ]),
```
I think `oneOf` is different than `oneofType`. `controlTypes` are string values, but `PropTypes.func` or `PropTypes.func.isRequired` is type definition.

villebro replied:
Whoops, my bad, didn't notice that.

rusackas commented on superset-frontend/src/explore/components/Control.jsx:
```diff
@@ -70,16 +75,18 @@ export default class Control extends React.PureComponent {
-    if (!this.props.type) return null; // this catches things like <hr/> elements (not a control!) shoved into the control panel configs.
-    const ControlType = controlMap[this.props.type];
+    const { type: controlType } = this.props;
+    if (!controlType) return null; // this catches things like <hr/> elements (not a control!) shoved into the control panel configs.
```
Side note, those really rub me the wrong way. At some point, I hope to open a PR that provides an alternate solution to that.

ktmud replied:
I think this comment is inaccurate. The JSX elements and nulls should be handled by [ControlPanelContainer.jsx](https://github.com/ktmud/incubator-superset/blob/efd0314819b91ebc00af1cbd4e32cb0b997d40e1/superset-frontend/src/explore/components/ControlPanelsContainer.jsx#L132-L137) already.

rusackas commented on superset-frontend/src/visualizations/presets/MainPreset.js:
```diff
@@ -39,7 +39,7 @@ import PivotTableChartPlugin from '@superset-ui/legacy-plugin-chart-pivot-table'
 import SankeyChartPlugin from '@superset-ui/legacy-plugin-chart-sankey';
 import SunburstChartPlugin from '@superset-ui/legacy-plugin-chart-sunburst';
-import TableChartPlugin from '@superset-ui/legacy-plugin-chart-table';
+import TableChartPlugin from '@superset-ui/plugin-chart-table';
```
🎉

rusackas commented on superset/viz.py:
```diff
@@ -572,6 +574,51 @@ class TableViz(BaseViz):
+        ):
+            raise QueryObjectValidationError(
+                _(
+                    "Choose either fields to [Group By] and [Metrics] and/or "
```
Found this error message a little confusing. Maybe something like:

`You cannot use [Columns] controls simultaneously with a combination of [Group By]/[Metrics]/[Percentage Metrics] controls. Please choose one or the other.`

ktmud replied:
Updated. P.S. This is for legacy API, most users would probably not see this anyway.

ktmud commented:
Github heads are messed up because of rebase. Close and reopen to see if it fixes CI.

rusackas reviewed (approved):
LGTM!

ktmud committed:
Run `npm dedupe` to simplify npm-lock.json

ktmud commented:
@nytai I've also [upgraded](https://github.com/apache/incubator-superset/pull/10113/commits/149286cc3520be34a8d9e335f1d0b56cb7390562) `react-table` to `v7.2.1` because that's what the new table chart is using. The only change related to your ListView work is that the `sortable` prop for columns is now called `canSort`.

ktmud commented:
Had to run `npm dedupe` to remove duplicate `react-table` in `package-lock.json`, but it seems to have broken jsdom. Will push a fix soon.

ktmud committed:
Upgrade jest and jsdom

villebro commented:
@ktmud I tried restarting this a few times, but it appears some unit tests are failing. Cypress seems to be fine, despite being red now (flaky test).

ktmud commented:
> @ktmud I tried restarting this a few times, but it appears some unit tests are failing. Cypress seems to be fine, despite being red now (flaky test).

Rebased and fixed another issue with js-dom upgrade and now it seems fine?

ktmud committed:
Disable a buggy a11y linting rule

ktmud committed:
Run npm dedupe && npm ci

ktmud committed:
try to fix a flaky test
